adam/perception/perception_graph_predicates.py

```diff
@@ -0,0 +1,644 @@
+            return False
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class CrossSectionPredicate(NodePredicate):
+    """
+    Represents constraints on a `Geon` given in a `PerceptionGraphPattern`
+    """
+
+    cross_section: CrossSection = attrib(validator=instance_of(CrossSection))
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+
+        unwrapped_graph_node = unwrap_if_necessary(graph_node)
+
+        if isinstance(unwrapped_graph_node, CrossSection):
+            return self.cross_section == unwrapped_graph_node
+        else:
+            return False
+
+    def dot_label(self) -> str:
+        return f"cross-section({self.cross_section})"
+
+    @staticmethod
+    def exactly_matching(cs: CrossSection) -> "CrossSectionPredicate":
+        return CrossSectionPredicate(cs)
+
+    def is_equivalent(self, other) -> bool:
+        if isinstance(other, CrossSectionPredicate):
+            return (
+                self.cross_section.curved == other.cross_section.curved
+                and self.cross_section.has_reflective_symmetry
+                == other.cross_section.has_reflective_symmetry
+                and self.cross_section.has_rotational_symmetry
+                == other.cross_section.has_rotational_symmetry
+            )
+        return False
+
+    def matches_predicate(self, predicate_node: "NodePredicate") -> bool:
+        if isinstance(predicate_node, CrossSectionPredicate):
+            return self.cross_section == predicate_node.cross_section
+        else:
+            return False
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class RegionPredicate(NodePredicate):
+    """
+    Represents constraints on a `Region` given in a `PerceptionGraphPattern`.
+    """
+
+    distance: Optional[Distance] = attrib(validator=optional(instance_of(Distance)))
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        # regions might be wrapped in tuples with their id()
+        # in order to simulate comparison by object ID.
+        unwrapped_graph_node = unwrap_if_necessary(graph_node)
+        return (
+            isinstance(unwrapped_graph_node, Region)
+            and self.distance == unwrapped_graph_node.distance
+        )
+
+    def dot_label(self) -> str:
+        return f"dist({self.distance})"
+
+    @staticmethod
+    def matching_distance(region: Region[Any]) -> "RegionPredicate":
+        return RegionPredicate(region.distance)
+
+    def is_equivalent(self, other) -> bool:
+        if isinstance(other, RegionPredicate):
+            if self.distance is None and other.distance is None:
+                return True
+            elif self.distance is not None and other.distance is not None:
+                return self.distance.name == other.distance.name
+        return False
+
+    def matches_predicate(self, predicate_node: "NodePredicate") -> bool:
+        if isinstance(predicate_node, RegionPredicate):
+            return predicate_node.distance == self.distance
```

Comment:
readability note, self. was listed as first argument in most comparisons in this file, but here predicate_node. is first.